service/dap/server.go

```diff
@@ -0,0 +1,464 @@
+	// or lines that do not have statements) will be skipped.
+	i := 0
+	for _, b := range request.Arguments.Breakpoints {
+		bp, err := s.debugger.CreateBreakpoint(
+			&api.Breakpoint{File: request.Arguments.Source.Path, Line: b.Line})
+		if err != nil {
+			s.log.Error("ERROR:", err)
+			continue
+		}
+		response.Body.Breakpoints[i].Verified = true
+		response.Body.Breakpoints[i].Line = bp.Line
+		i++
+	}
+	response.Body.Breakpoints = response.Body.Breakpoints[:i]
+	s.send(response)
+}
+
+func (s *Server) onSetExceptionBreakpointsRequest(request *dap.SetExceptionBreakpointsRequest) {
+	// Unlike what DAP documentation claims, this request is always sent
+	// even though we specified no filters at initializatin. Handle as no-op.
+	s.send(&dap.SetExceptionBreakpointsResponse{Response: *newResponse(request.Request)})
+}
+
+func (s *Server) onConfigurationDoneRequest(request *dap.ConfigurationDoneRequest) {
+	if s.stopOnEntry {
+		e := &dap.StoppedEvent{
+			Event: *newEvent("stopped"),
+			Body:  dap.StoppedEventBody{Reason: "breakpoint", ThreadId: 1, AllThreadsStopped: true},
+		}
+		s.send(e)
+	}
+	s.send(&dap.ConfigurationDoneResponse{Response: *newResponse(request.Request)})
+	if !s.stopOnEntry {
+		s.doContinue()
+	}
+}
+
+func (s *Server) onContinueRequest(request *dap.ContinueRequest) {
+	s.send(&dap.ContinueResponse{Response: *newResponse(request.Request)})
+	s.doContinue()
+}
+
+func (s *Server) sendErrorResponse(request dap.Request, id int, summary string, details string) {
+	er := &dap.ErrorResponse{}
+	er.Type = "response"
+	er.Command = request.Command
+	er.RequestSeq = request.Seq
+	er.Success = false
+	er.Message = summary
+	er.Body.Error.Id = id
+	er.Body.Error.Format = fmt.Sprintf("%s: %s", summary, details)
+	s.log.Error(er.Body.Error.Format)
+	s.send(er)
+}
+
+func (s *Server) sendUnsupportedErrorResponse(request dap.Request) {
+	s.sendErrorResponse(request, 9999, "Unsupported command",
+		fmt.Sprintf("cannot process '%s' request", request.Command))
+}
+
+func newResponse(request dap.Request) *dap.Response {
+	return &dap.Response{
+		ProtocolMessage: dap.ProtocolMessage{
+			Seq:  0,
+			Type: "response",
+		},
+		Command:    request.Command,
+		RequestSeq: request.Seq,
+		Success:    true,
+	}
+}
+
+func newEvent(event string) *dap.Event {
+	return &dap.Event{
+		ProtocolMessage: dap.ProtocolMessage{
+			Seq:  0,
+			Type: "event",
+		},
+		Event: event,
+	}
+}
+
+func (s *Server) doContinue() {
+	if s.debugger == nil {
+		return
+	}
+	state, err := s.debugger.Command(&api.DebuggerCommand{Name: api.Continue})
+	if err != nil {
+		s.log.Error(err)
+		switch err.(type) {
+		case proc.ErrProcessExited:
+			e := &dap.TerminatedEvent{Event: *newEvent("terminated")}
+			s.send(e)
+		default:
+		}
+		return
+	}
+	if state.Exited {
+		e := &dap.TerminatedEvent{Event: *newEvent("terminated")}
+		s.send(e)
+	} else {
+		e := &dap.StoppedEvent{Event: *newEvent("stopped")}
+		e.Body.Reason = "breakpoint"
```

Comment:
This could also be `pause` if a halt command was sent. I don't know if it makes a difference and it will probably be hard to make this distinction (so if it doesn't matter in practice it probably isn't worth bothering with this).

Reply:
There is such thing as a PauseRequest, but I have not implemented that in this PR, so no checking for this case here yet. I don't think I can tell from the debugger if this was a pause or breakpoint, can I? I guess I could keep track if this was called in a context of a breakpoint or a halt request.

Reply:
> I don't think I can tell from the debugger if this was a pause or breakpoint, can I? I guess I could keep track if this was called in a context of a breakpoint or a halt request.

Yes, you would have to keep track (and there would be a race condition if the user requested a pause just as the program hit a breakpoint).